Invoke-WebRequest and Invoke-RestMethod: Rename TimeoutSec to ConnectionTimeoutSeconds (with alias) and add OperationTimeoutSeconds

stevenebutler opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

This renames `-TimeoutSec` parameter (with alias to old parameter name) to `-ConnectionTimeoutSeconds` and adds `-OperationTimeoutSeconds` parameter to both `Invoke-WebRequest` and `Invoke-RestMethod` commands.

* `-TimeoutSec` becomes `-ConnectionTimeoutSeconds` and is aliased with `-TimeoutSec` and has same behaviour as before. That is, it is the timeout that applies from just prior to sending a web request to receiving the response headers.
*  `-OperationTimeoutSeconds` specifies the maximum allowed time between reads from the network when streaming http response content.

The default for both items is to have no timeout.

close #12249
close #16122 
close #19519

## PR Context

There are reports that when a network stream goes away and the OS doesn't or is unable to notify PowerShell that the network stream has been lost that PowerShell hangs forever in `Invoke-WebRequest` / `Invoke-RestMethod` on stalled response streams.

A previous PR made it possible to use CTRL-C to terminate this, but there was no way to have PowerShell timeout after a period of network inactivity in batch scenarios. This PR adds the timeout. The timeout applies to inter-stream reads, and not to the stream time as a whole. Therefore setting the `-OperationTimeoutSeconds` to 30 seconds means that any delay of longer than 30 seconds between reading data from the stream will terminate the request. However a large file that takes several minutes to download will not terminate unless the stream stalls for more than 30 seconds.

The default of 0 for `-ConnectionTimeoutSeconds` and `-OperationTimeoutSeconds` means no timeout applies as per request by WG noted in #19519. Any value less than 0 for `-OperationTimeoutSeconds` is also treated as no timeout.

### Performance Comparison

The following benchmark was used to assess performance implications of this PR over 7.3.4 vanilla using a local dotnet minimal API with a static array of bytes being returned by the website (source below).

All tests were run on a HP laptop with 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz   2.70 GHz and 16 GB memory. Note the dotnet web server and the client application were tested over HTTP on the same laptop over localhost interface.

Note: testing was done on an earlier iteration before renaming and minor refactoring of timeout handling.

```csharp
var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();

var bytes = new byte[200*1024*1024]; // ~200 MB array
Random.Shared.NextBytes(bytes);
Console.WriteLine("Byte stream initialized");
app.MapGet("/", () => Results.Bytes(bytes));

app.Run();
```
The benchmark web server was built in Release configuration and run as an executable.

```powershell
# Curl from 7.3.4 vanilla
1 .. 10 | %{(Measure-Command { curl http://localhost:5000/ --output curl_data.bin }).TotalMilliseconds} | Measure-Object -Average -Minimum -Maximum -StandardDeviation
# Command run twice, and second summary taken
Count             : 10
Average           : 733.3488
Maximum           : 879.6177
Minimum           : 632.9115
StandardDeviation : 69.7117181472224

# PowerShell 7.3.4 vanilla
1 .. 10 | %{(Measure-Command { iwr http://localhost:5000/ -OutFile iwr_7.3.4_data.bin }).TotalMilliseconds} | Measure-Object -Average -Minimum -Maximum -StandardDeviation

# Command run twice and second summary taken
Count             : 10
Average           : 1115.43297
Sum               :
Maximum           : 1132.2394
Minimum           : 1102.5451
StandardDeviation : 11.4286089899233

# Powershell 7.4.0-preview.3
1 .. 10 | %{(Measure-Command { iwr http://localhost:5000/ -OutFile iwr_7.4.0-preview-3_data.bin }).TotalMilliseconds} | Measure-Object -Average -Minimum -Maximum -StandardDeviation

# Command run twice and second summary taken 
Count             : 10
Average           : 415.06509
Sum               :
Maximum           : 474.9918
Minimum           : 388.273
StandardDeviation : 33.2625053366048
Property          :

# Powershell 7.4.0-preview.3-81-g6b80bf03b74dff76816740fbb3549b87f190f518
1 .. 10 | %{(Measure-Command { iwr http://localhost:5000/ -OutFile iwr_7.4.0-preview.3-with-19558_data.bin }).TotalMilliseconds} | Measure-Object -Average -Minimum -Maximum -StandardDeviation

# Command run twice and second summary taken
Count             : 10
Average           : 410.43762
Sum               :
Maximum           : 479.3415
Minimum           : 378.843
StandardDeviation : 27.3532692743502
Property          :
# Powershell 7.4.0-preview.3-81-g6b80bf03b74dff76816740fbb3549b87f190f518 - Adding OperationTimeoutSeconds parameter
1 .. 10 | %{(Measure-Command { iwr http://localhost:5000/ -OperationTimeoutSeconds 10 -OutFile iwr_7.4.0-preview.3-with-19558-OperationTimeoutSeconds-data.bin }).TotalMilliseconds} | Measure-Object -Average -Minimum -Maximum -StandardDeviation

# Command run twice and second summary taken
Count             : 10
Average           : 646.04633
Sum               :
Maximum           : 663.085
Minimum           : 607.9275
StandardDeviation : 16.2551572417324
Property          :
```

My conclusions from these results are:

*  the patch does not have any material impact on performance on downloading a file from the last preview release when the `-OperationTimeoutSeconds` parameter isn't used
* the patch has a 30% increase on download performance over release 3 when the `-OperationTimeoutSeconds` parameter is used but is still favourable when compared to curl performance. This is likely attributable to the work spent resetting the timeout on the cancellation token before each network read.
* the last preview release is significantly faster than 7.3.4 for this use case as it is now faster than using `curl` when it was significantly slower in the 7.3.4 release.


## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

CarloToso commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -240,7 +240,7 @@ private HttpClient CreateHttpClient()
             return new HttpClient(handler)
             {
-                Timeout = _timeoutSec is 0 ? TimeSpan.FromMilliseconds(Timeout.Infinite) : TimeSpan.FromSeconds(_timeoutSec)
+                Timeout = _connectTimeoutSec is 0 ? TimeSpan.FromMilliseconds(Timeout.Infinite) : TimeSpan.FromSeconds(_connectTimeoutSec)
```
```suggestion
                Timeout = _connectTimeoutSec > 0 ? TimeSpan.FromSeconds(_connectTimeoutSec) : Timeout.InfiniteTimeSpan
```

CarloToso commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/StreamHelper.cs:
```diff
@@ -253,6 +257,59 @@ private void Initialize(CancellationToken cancellationToken = default)
 
+    internal static class StreamTimeoutExtensions
+    {
+        internal static async Task<int> ReadAsync(this Stream stream, Memory<byte> buffer, TimeSpan readTimeout, CancellationToken cancellationToken)
```
Why didn't you use perReadTimeout here?

stevenebutler replied:
There is only one read here so it doesn't make sense to call it per read timeout.

CarloToso replied:
Maybe you could just always use `readTimeout`

stevenebutler replied:
I think it's clearer that the read timeout applies to every read in the copyToAsync with the current name rather than the copy as a whole.

CarloToso commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs:
```diff
@@ -72,24 +72,36 @@ public class WebResponseObject
+        /// <summary>
+        /// Time permitted between reads or Timeout.InfiniteTimeSpan for no timeout.
+        /// </summary>
+        protected TimeSpan perReadMillisecondsTimeout;
```
@stevenebutler why did you choose `perReadMillisecondsTimeout` instead of something like `_perReadTimeout` ?
Maybe you can cache `perReadTimeout = NetworkTimeout` directly in `WebRequestPSCmdlet.Common.cs`

stevenebutler replied:
Even if cached in the cmdlet, it would still need to be passed down into this class because it needs to know the value.

I used a getter to convert the value in the cmdlet and came locally when used multiple times but could change it to a pre calculated value if you think it's worthwhile. I did consider this but didn't think it would be of much benefit.

stevenebutler replied:
I have renamed the protected property to perReadTimeout but I won't cache the TimeSpan value in the web request unless someone feels strongly this should be done.

stevenebutler commented:
Reopening - accidentally force-pushed empty commits after a squash merge.

iSazonov commented:
1. I think this PR is wrong. There's a fundamental problem there https://github.com/dotnet/runtime/issues/74568#issuecomment-1238414504
It should be solved in .Net, rather than creating a lot of useless code in PowerShell.

2. That issue also clearly explains why the NetworkTimeout parameter makes no sense at all.

3. Such an unnecessary complication https://github.com/PowerShell/PowerShell/pull/19330 This should be rolled back. The same thing can be done in a much simpler way. Not only is the code unnecessarily complicated, but new bugs seem to have been created.

I urge MSFT team to look at this carefully.

stevenebutler commented:
@iSazonov - if the previous PR#19330 caused bugs then please provide the issue numbers for those bugs as I was not aware of any issues with the fix. There were some races in the tests for the feature which are difficult to avoid when doing tests that require timing of events when run in a busy CI system that may not respond as quickly as you'd normally expect. The tests initially had a tolerance of a few 100 ms but this was too low for the CI agents, so I have used longer time ranges for timing related tests. If the tests continue to be problematic, they could be disabled, but to the best of my knowledge the fundamental issue with CTRL-C not always working (when the connection was stalled) was resolved by that PR. If you think otherwise then please provide some concrete evidence in the form of a non-working test case and I will investigate.

I have implemented this PR in accordance with advice provided by Microsoft on the issue you linked in your comment on how to achieve the result and the previous PR laid the groundwork for supporting cancellation throughout the command.

@TravisEz13 - are you requesting that I as PR author perform that test or will someone within MSFT be doing that as part of the release testing process?

I didn't think this would be considered a breaking change since it provides an alias for the original parameter so scripts should continue to work. It may be breaking for users who instantiate the cmdlet class and manipulate the parameter directly though.

stevenebutler commented:
@TravisEz13 - I have added some benchmarks and methodology for those benchmarks to the summary of the PR testing with a 200 MB randomly generated byte sequence across curl, pwsh 7.3.4, pwsh 7.4.0-preview3 and the latest iteration of this PR.

daxian-dbw commented:
@stevenebutler The signature of public constructor `BasicHtmlWebResponseObject` is changed, both in this PR and https://github.com/PowerShell/PowerShell/pull/19330, which is a breaking change. We discussed about it and thought this breaking change falls in the [Bucket 3](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/breaking-change-contract.md#bucket-3-unlikely-grey-area), and decide to accept it for now and wait on feedback.

@iSazonov It will make things easier when https://github.com/dotnet/runtime/issues/75188 is addressed. However, it's in `Future` milestone and thus it's unknown when it will be implemented in .NET. So, #19330 and this PR seem to be needed if we want the targeted PowerShell issues to be addressed now. I will let the Cmdlet WG to review this issue.

stevenebutler commented:
> @stevenebutler The signature of public constructor `BasicHtmlWebResponseObject` is changed, both in this PR and #19330, which is a breaking change. We discussed about it and thought this breaking change falls in the [Bucket 3](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/breaking-change-contract.md#bucket-3-unlikely-grey-area), and decide to accept it for now and wait on feedback.

Thanks @daxian-dbw  - that makes sense. I was only thinking about it in terms of scripting users.

iSazonov commented:
> @iSazonov It will make things easier when [dotnet/runtime#75188](https://github.com/dotnet/runtime/issues/75188) is addressed. However, it's in `Future` milestone and thus it's unknown when it will be implemented in .NET. So, #19330 and this PR seem to be needed if we want the targeted PowerShell issues to be addressed now. I will let the Cmdlet WG to review this issue.

@daxian-dbw [dotnet/runtime#75188](https://github.com/dotnet/runtime/issues/75188) replaces my original issue where their team was very skeptical. So I don't think they're going to do anything about it at all unless of course the MSFT PowerShell team asks them to. I assume they are coming from the experience of typical services both server and client, where sessions are just restarted by the application itself. (In other words there is a problem but it is hidden.) We're referring to a slightly different scenario where blocking the main thread is unacceptable. I think root issue is important to be fixed, because it forces developers to create complexities in their code with no guarantee of correctness and achievability of the result.
A typical example is #19330 which does not solve all problems and moreover seems to add bugs. 
I would prefer to roll this back and if we really want to fix Ctrl-C, just offload the code to a new thread with `Task.Run()`. If [dotnet/runtime#75188](https://github.com/dotnet/runtime/issues/75188) is implemented, we probably won't need to change anything else.

stevenebutler commented:
> A typical example is #19330 which does not solve all problems and moreover seems to add bugs.

You keep saying PR#19330 seems to add bugs but you have not actually provided or linked to an existing bug report that is attributable to that PR. If you are aware of a bug caused by that PR please let me know so I can look at it.

> I would prefer to roll this back and if we really want to fix Ctrl-C, just offload the code to a new thread with Task.Run(). If https://github.com/dotnet/runtime/issues/75188 is implemented, we probably won't need to change anything else.

While offloading the stuck activity to a new task with Task.Run() is simpler, in reality it's still stuck and is now consuming resources that may never be freed. The cancellation token approach allows the stuck activity to actually be cancelled and gives a chance for resources to be freed.

SteveL-MSFT commented:
queued for WG discussion

stevenebutler committed:
Rename TimeoutSec to ConnectionTimeoutSeconds and add OperationTimeoutSeconds

TimeoutSec becomes ConnectionTimeoutSeconds and has same behaviour as before.
OperationTimeoutSeconds specifies the maximum allowed time between reads from the
network when streaming http response messages.

The default for both items is to have no timeout.

Co-Authored-By: CarloToso <105941898+CarloToso@users.noreply.github.com>

SteveL-MSFT commented:
@PowerShell/wg-powershell-cmdlets reviewed this, we agree with the added functionality.  However, we suggest renaming `TimeoutSec` to `ConnectionTimeoutSeconds` and the new parameter `OperationTimeoutSeconds` as "NetworkTimeoutSec" can still be a bit ambiguous what it is referring to.  We also prefer to spell out "Seconds" since the parameter name is already verbose and could be confused with "Security".

stevenebutler commented:
Ok, I have finished updating as per recommendations. I also did some minor refactoring of the timeout calculation code to share the conversion from integer seconds to a TimeSpan between the two timeouts.

SteveL-MSFT reviewed (changes requested):
Great change!  Overall, looks good, but one comment on the fully qualified error id.

stevenebutler commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -608,7 +608,15 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+        $command = "Invoke-WebRequest -Uri '$uri' -ConnectionTimeoutSeconds 2"
+
+        $result = ExecuteWebCommand -command $command
+        $result.Error.FullyQualifiedErrorId | Should -Be "System.Threading.Tasks.TaskCanceledException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand"
```
@SteveL-MSFT do you know of any examples of cmdlets that do something similar?

Wouldn't changing the exception reported for ConnectionTimeoutReached be a potential for a script breaking change as the existing behaviour has been around now for some time and could be caught by existing scripts' try/catch blocks?

stevenebutler replied:
@SteveL-MSFT I have updated the code/tests to set the FullyQualifiedErrorId as per your requested change.

stevenebutler committed:
Merge remote-tracking branch 'origin/master' into Rename-TimeoutSec-to-ConnectTimeoutSec-and-alias-to-TimeoutSec-and-add-NetworkTimeoutSec-timeout

stevenebutler committed:
Set FullyQualifiedErrorId as per feedback from review

stevenebutler committed:
Merge branch 'PowerShell:master' into Rename-TimeoutSec-to-ConnectTimeoutSec-and-alias-to-TimeoutSec-and-add-NetworkTimeoutSec-timeout

daxian-dbw commented:
@stevenebutler Thanks for your contribution!
